rust/relay/ebpf-turn-router/src/move_payload.rs

```diff
@@ -0,0 +1,149 @@
+use crate::{channel_data::CdHdr, error::Error};
+use aya_ebpf::{helpers::bpf_xdp_adjust_tail, programs::XdpContext};
+use network_types::{
+    eth::EthHdr,
+    ip::{Ipv4Hdr, Ipv6Hdr},
+    udp::UdpHdr,
+};
+
+// Set an upper limit for bounds checks
+const MAX_MTU: usize = 1500;
+const MAX_PAYLOAD: usize = MAX_MTU - EthHdr::LEN - Ipv4Hdr::LEN - UdpHdr::LEN - CdHdr::LEN;
+
+#[inline(always)]
+pub fn add_channel_data_header_ipv4(ctx: &XdpContext, header: CdHdr) -> Result<(), Error> {
+    adjust_header::<{ Ipv4Hdr::LEN }>(ctx, Some(&header))
+}
+
+#[inline(always)]
+pub fn add_channel_data_header_ipv6(ctx: &XdpContext, header: CdHdr) -> Result<(), Error> {
+    adjust_header::<{ Ipv6Hdr::LEN }>(ctx, Some(&header))
+}
+
+#[inline(always)]
+pub fn remove_channel_data_header_ipv4(ctx: &XdpContext) -> Result<(), Error> {
+    adjust_header::<{ Ipv4Hdr::LEN }>(ctx, None)
+}
+
+#[inline(always)]
+pub fn remove_channel_data_header_ipv6(ctx: &XdpContext) -> Result<(), Error> {
+    adjust_header::<{ Ipv6Hdr::LEN }>(ctx, None)
+}
+
+/// Add or remove channel data header
+#[inline(never)]
+fn adjust_header<const IP_HEADER_LEN: usize>(
+    ctx: &XdpContext,
+    header: Option<&CdHdr>,
+) -> Result<(), Error> {
+    let offset = EthHdr::LEN + IP_HEADER_LEN + UdpHdr::LEN;
+
+    if let Some(hdr) = header {
+        let ret = unsafe { bpf_xdp_adjust_tail(ctx.ctx, CdHdr::LEN as i32) };
+        if ret < 0 {
+            return Err(Error::XdpAdjustTailFailed(ret));
+        }
+
+        let data_start = ctx.data();
+        let data_end = ctx.data_end();
+
+        copy_bytes(data_start, data_end, offset, true);
+
+        // Copy header
+        let hdr_dst = data_start + offset;
+        let hdr_src = hdr as *const CdHdr as *const u8;
+
+        for i in 0..CdHdr::LEN {
+            if hdr_dst + i < data_end {
+                let dst_ptr = (hdr_dst + i) as *mut u8;
+                unsafe {
+                    *dst_ptr = *hdr_src.add(i);
+                }
+            }
+        }
+    } else {
+        // Remove channel data header
+        let data_start = ctx.data();
+        let data_end = ctx.data_end();
+
+        copy_bytes(data_start, data_end, offset, false);
+
+        let ret = unsafe { bpf_xdp_adjust_tail(ctx.ctx, -(CdHdr::LEN as i32)) };
+        if ret < 0 {
+            return Err(Error::XdpAdjustTailFailed(ret));
+        }
+    }
+
+    Ok(())
+}
+
+/// Copy bytes from src_offset to dst_offset being mindful of overlap.
```

Comment:
Perhaps this kind of docs / knowledge could go on the module (top of the file using `!//`) as it sort of applies to the entire approach used here. If you agree with splitting up the fn, it also means you don't need to duplicate it :)